internal/status/immediatestatus.go

```diff
@@ -0,0 +1,138 @@
+type StatusObserver struct {
+	// goalStateEventMap is a map that stores the goal state key and the status item
+	// sync.Map is preferred over map because it is safe for concurrent use
+	goalStateEventMap sync.Map
+
+	// ctx is the logger context
+	ctx *log.Context
+
+	// Reporter is the status Reporter
+	Reporter statusreporter.IGuestInformationServiceClient
+}
+
+func (o *StatusObserver) Initialize(ctx *log.Context) {
+	o.goalStateEventMap = sync.Map{}
+	o.ctx = ctx
+	o.Reporter = statusreporter.NewGuestInformationServiceClient(hostgacommunicator.WireServerFallbackAddress)
+}
+
+func (o *StatusObserver) OnNotify(status types.StatusEventArgs) error {
+	o.ctx.Log("message", fmt.Sprintf("Processing status event for goal state with key %v", status.StatusKey))
+	o.goalStateEventMap.Store(status.StatusKey, status.TopLevelStatus)
+	return o.reportImmediateStatus(o.getImmediateTopLevelStatusToReport())
+}
+
+func (o *StatusObserver) getImmediateTopLevelStatusToReport() ImmediateTopLevelStatus {
+	o.ctx.Log("message", "Getting all goal states from the event map with the latest status that are not empty")
+	latestStatusToReport := []ImmediateStatus{}
+	o.goalStateEventMap.Range(func(key, value interface{}) bool {
+		// Only report the latest active status for each goal state
+		if value.(types.StatusItem) != (types.StatusItem{}) {
+			statusItem := value.(types.StatusItem)
+			immediateStatus := ImmediateStatus{
+				SequenceNumber: key.(types.GoalStateKey).SeqNumber,
+				TimestampUTC:   statusItem.TimestampUTC,
+				Status:         statusItem,
+			}
+			latestStatusToReport = append(latestStatusToReport, immediateStatus)
+		}
+		return true
+	})
+
+	o.ctx.Log("message", "Creating immediate status to report")
+	return ImmediateTopLevelStatus{
+		AggregateHandlerImmediateStatus: []ImmediateHandlerStatus{
+			{
+				HandlerName:              "RunCommandHandler",
+				AggregateImmediateStatus: latestStatusToReport,
+			},
+		},
+	}
+}
+
+func (o *StatusObserver) reportImmediateStatus(immediateStatus ImmediateTopLevelStatus) error {
+	o.ctx.Log("message", "Marshalling immediate status into json")
+	rootStatusJson, err := json.Marshal(immediateStatus)
+	if err != nil {
+		return fmt.Errorf("status: failed to marshal immediate status into json: %v", err)
+	}
+
+	o.ctx.Log("message", "create request to upload status to: "+o.Reporter.GetPutStatusUri())
+	response, err := o.Reporter.ReportStatus(string(rootStatusJson))
+	if err != nil {
+		return errors.Wrap(err, "failed to report status to HGAP")
+	}
+
+	if response.StatusCode != 200 {
+		return errors.New("failed to report status with error code " + response.Status)
+	}
+
+	o.ctx.Log("message", fmt.Sprintf("Status received from request to %v: %v", response.Request.URL, response.Status))
+	return nil
+}
+
+// Remove the goal states that have already been processed from the event map
+// If the goal state that was added before is not in the new list of goal states, it should be removed
```

Comment:
Eventually we'll need to report these too, but for now this is ok.

Reply:
This will be a subsequent item, but we need to hold onto the status for all RunCommands previously processed and send them in the status. That way CRP can just retrieve the latest status for all commands.
When a command is sent down with toBeDeleted = true, then the status document may be deleted and the extension should no longer report on it.

Reply:
I see, makes sense. Thanks for the explanation!